virttest/utils_test/qemu.py

```diff
@@ -1470,6 +1472,96 @@ def __init__(self, test, params, env):
+        :return mem_size: total physical memory in guest.
+        """
+        error_context.context("Get physical memory in guest")
+        mem_size = 0.0
+        session = self.get_session(vm)
+        if self.params.get("os_type") == "windows":
+            cmd = 'systeminfo | findstr /C:"Total Physical Memory"'
+            regex = re.compile("Total Physical Memory:\s+(.*)", re.M | re.I)
+        else:
+            cmd = 'cat /proc/meminfo'
+            regex = re.compile("memtotal:\s+(.*)", re.M | re.I)
+        output = session.cmd_output_safe(cmd, timeout=240)
+        try:
+            mem_str = regex.search(output.replace(',', '')).groups()[0]
+            args = ("%sB" % mem_str, "M", 1024)
+            mem_size = float(utils_misc.normalize_data_size(*args))
+            logging.info("total memory reported by os: %.2f MB" % mem_size)
+        except Exception:
+            logging.warn("invalid output from guest: %s" % output)
+        return mem_size
+
+    @error_context.context_aware
+    def get_guest_free_mem(self, vm):
+        """
+        get free memory size detect by guest os.
+
+        :param vm: vm object.
+        :return mem_size: free physical memory in guest.
+        """
+        error_context.context("Get free memory in guest")
+        mem_size = 0.0
+        session = self.get_session(vm)
+        if self.params.get("os_type") == "windows":
+            cmd = 'systeminfo | findstr /C:"Available Physical Memory"'
+            regex = re.compile("Available Physical Memory:\s+(.*)", re.M | re.I)
+        else:
+            cmd = 'cat /proc/meminfo'
+            regex = re.compile("memfree:\s+(.*)", re.M | re.I)
+        output = session.cmd_output_safe(cmd, timeout=240)
+        try:
+            mem_str = regex.search(output.replace(',', '')).groups()[0]
+            args = ("%sB" % mem_str, "M", 1024)
+            mem_size = float(utils_misc.normalize_data_size(*args))
+            logging.info("free memory reported by os: %.2f MB" % mem_size)
+        except Exception:
+            logging.warn("invalid output from guest: %s" % output)
+        return mem_size
+
+    def get_session(self, vm):
+        """
+        Get connection for VM
+        """
+        key = vm.instance
+        if not self.sessions.get(key):
+            self.sessions[key] = []
+        else:
+            for session in self.sessions[key]:
+                if session.is_responsive():
+                    return session
+                else:
+                    session.close()
+                    self.sessions[key].remove(session)
+        login_timeout = float(self.params.get("login_timeout", 600))
+        session = vm.wait_for_login(timeout=login_timeout)
+        self.sessions[key].append(session)
+        return session
+
+    def close_sessions(self):
+        """
+        Close opening session, better to call it in the end of test.
+        """
+        while self.sessions:
+            _, sessions = self.sessions.popitem()
```

Comment:
for session in self.sessions.popitem()[1]:
      session.close()
Looks more beautiful .